README.md
# virtual-audio-graph

[![npm version](https://badge.fury.io/js/virtual-audio-graph.svg)](http://badge.fury.io/js/virtual-audio-graph)
![CI/CD](https://github.com/benji6/virtual-audio-graph/workflows/CI/CD/badge.svg)
![NPM](https://img.shields.io/npm/l/virtual-audio-graph)
[![Netlify Status](https://api.netlify.com/api/v1/badges/f5f3984d-617b-418a-9c5b-8881aceb7adb/deploy-status)](https://app.netlify.com/sites/virtual-audio-graph/deploys)

## Overview

Small and dependency-free library for declaratively manipulating the Web Audio API.

virtual-audio-graph manages the state of the audio graph so this does not have to be done manually. Simply declare what you would like the audio graph to look like and virtual-audio-graph takes care of the rest.

Inspired by [virtual-dom](https://github.com/Matt-Esch/virtual-dom) and [React](https://github.com/facebook/react).

## Installation

```bash
npm i -S virtual-audio-graph
```

virtual-audio-graph is distributed as a bundled CJS module by default, however, there is also a `module` property in its `package.json` so tools like [Rollup](https://github.com/rollup/rollup) and [Webpack](https://github.com/webpack/webpack) can consume an ES modules build.

## Size

virtual-audio-graph is designed to be small and weighs in at 11.9kB minified (3.3kB minified & gzipped) according to [bundlephobia](https://bundlephobia.com) (https://bundlephobia.com/result?p=virtual-audio-graph@1.1.8).

## Docs

**[First check out the virtual-audio-graph guide](https://virtual-audio-graph.netlify.com)** for working examples and to understand how the library works.

Then see the [API docs for all supported virtual audio node factory functions](docs/standard-nodes.md).

For an example of virtual-audio-graph working in a real application you can take a look at [Andromeda](https://github.com/benji6/andromeda) which is one of my other projects.


CHANGELOG.md
# Release Notes

## **This Changelog is no longer in use, please see GitHub releases for changes made since 0.19.x.**

## 0.19.x

### Added

`bufferSource` now supports `offsetTime`

## 0.18.x

virtual-audio-graph no longer has any external dependencies

## 0.17.x

Breaking API change - `defineNodes` removed. `virtual-audio-graph` no longer internally remembers the custom node creator functions and these must be passed directly. This is more inline with how virtual-dom libraries handle component abstractions and keeps `virtual-audio-graph`'s internal state management to the audio graph alone

```javascript
// old API
virtualAudioGraph.defineNodes({ coolCustomNode });
virtualAudioGraph.update({ 0: ["coolCustomNode", "output", { someParams }] });
```

```javascript
// new API
virtualAudioGraph.update({ 0: [coolCustomNode, "output", { someParams }] });
```

## 0.16.x

Using [Rollup](https://github.com/rollup/rollup) for bundling ES5 dist.

Added `jsnext:main` ([more info](https://github.com/rollup/rollup/wiki/jsnext:main)) to package.json

## 0.15.x

Breaking API change - `defineNode` replaced by `defineNodes`:

```javascript
// old API
virtualAudioGraph.defineNode(coolEffect, "coolEffect");
virtualAudioGraph.defineNode(someOscillators, "someOscillators");
```

```javascript
// new API
virtualAudioGraph.defineNodes({ coolEffect, someOscillators });
```

## 0.14.x

Added support for AudioParam methods

## 0.13.x

`virtual-audio-graph` now exports a factory instead of a constructor

```javascript
// pre 0.13.x
import VirtualAudioGraph from "virtual-audio-graph";
const virtualAudioGraph = new VirtualAudioGraph();
```

```javascript
// 0.13.x
import createVirtualAudioGraph from "virtual-audio-graph";
const virtualAudioGraph = createVirtualAudioGraph();
```

Because of the way the new operator works this is not a breaking change

## 0.12.x

Breaking API change:

```javascript
const newAPI = {
  0: ["oscillator", "output"],
  1: ["gain", { key: 0, destination: "detune" }, { gain: 0.5 }, "input"],
  2: ["oscillator", 1, { frequency: 110 }],
};
```

```javascript
const oldAPI = {
  0: {
    node: "oscillator",
    output: "output",
  },
  1: {
    input: "input",
    node: "gain",
    output: { key: 0, destination: "detune" },
    params: {
      gain: 0.5,
    },
  },
  2: {
    node: "oscillator",
    output: 1,
    params: {
      frequency: 110,
    },
  },
};
```

---

## 0.11.x

- Added support for:
  - ChannelMergerNode
  - ChannelSplitterNode

---

## 0.10.x

- Added support for:
  - MediaStreamAudioDestinationNode
  - MediaStreamAudioSourceNode

---

## 0.9.x

- Added support for:
  - ConvolverNode
  - DynamicsCompressorNode
  - WaveShaperNode

---

## 0.8.x

- Added getAudioNodeById method to virtualAudioGraph

- Added support for:
  - AnalyserNode
  - AudioBufferSourceNode

---

## 0.7.x

Prior to version 0.7.x virtual-audio-graph parameters were an array of objects with id properties representing nodes like this:

```javascript
[
  {
    id: 0,
    node: "oscillator",
    output: "output",
    params: {
      frequency: 220,
    },
  },
  {
    id: 1,
    node: "oscillator",
    output: { id: 0, destination: "detune" },
    params: {
      frequency: 110,
    },
  },
];
```

Now the parameters are a single object with keys which represent the node ids:

```javascript
{
  0: {
    node: 'oscillator',
    output: 'output',
    params: {
      frequency: 220,
    },
  },
  1: {
    node: 'oscillator',
    output: {key: 0, destination: 'detune'}, // NB. "key" property used to be "id"
    params: {
      frequency: 110,
    },
  },
}
```

The new notation automatically ensures the id of each node exists and is unique. It is also more concise and allows for greater performance.


benchmark.js
require('web-audio-test-api')
/* global WebAudioTestAPI   */
/* eslint-disable no-console */
WebAudioTestAPI.setState('AudioContext#createStereoPanner', 'enabled')
const Benchmark = require('benchmark')
const createPublishedVirtualAudioGraph = require('virtual-audio-graph')
const createDevelopmentVirtualAudioGraph = require('./dist/index')
const pingPongDelay = require('./test/utils/pingPongDelay')

const runBenchmarkCode = function (virtualAudioGraph) {
  virtualAudioGraph.update({0: ['gain', 'output']})

  virtualAudioGraph.update({
    0: ['gain', 'output', {gain: 0.5}],
    1: ['oscillator', 0, {
      detune: 14,
      frequency: 1720,
      type: 'triangle',
    }],
    2: ['oscillator', 0, {
      detune: -2,
      frequency: 880,
      type: 'square',
    }],
    3: ['oscillator', 0, {
      detune: 1,
      frequency: 440,
      type: 'sine',
    }],
    4: ['oscillator', 0, {
      detune: -20,
      frequency: 220,
      type: 'sawtooth',
    }],
  })

  virtualAudioGraph.update({0: ['gain', 'output']})

  virtualAudioGraph.update({
    0: ['gain', 'output'],
    1: ['oscillator', {destination: 'gain', key: 0}, {frequency: 100}],
  })

  virtualAudioGraph.update({
    0: ['gain', 'output', {gain: 0.5}],
    1: [pingPongDelay, 0],
    2: ['oscillator', 1, {
      detune: 14,
      frequency: 1720,
      type: 'triangle',
    }],
    3: ['oscillator', 1, {
      detune: -2,
      frequency: 880,
      type: 'square',
    }],
    4: ['oscillator', 1, {
      detune: 1,
      frequency: 440,
      type: 'sine',
    }],
    5: ['oscillator', 1, {
      detune: -20,
      frequency: 220,
      type: 'sawtooth',
    }],
  })

  virtualAudioGraph.update({
    0: ['gain', 'output', {gain: 0.5}],
    1: [pingPongDelay, 0],
    2: ['oscillator', 1, {
      detune: 45,
      frequency: 123,
      type: 'triangle',
    }],
    3: ['oscillator', 1, {
      detune: -23,
      frequency: 123,
      type: 'square',
    }],
    4: ['oscillator', 1, {
      detune: 12,
      frequency: 123,
      type: 'sine',
    }],
    5: ['oscillator', 1, {
      detune: -260,
      frequency: 123,
      type: 'sawtooth',
    }],
  })

  virtualAudioGraph.update({0: ['gain', 'output']})
  virtualAudioGraph.update({})
}

new Benchmark.Suite()
  .add('PublishedVirtualAudioGraph', () => runBenchmarkCode(createPublishedVirtualAudioGraph()))
  .add('DevelopmentVirtualAudioGraph', () => runBenchmarkCode(createDevelopmentVirtualAudioGraph()))
  .on('cycle', event => console.log(String(event.target)))
  .on('complete', function () {
    console.log('Fastest is ' + this.filter('fastest').map('name'))
  })
  .run({async: true})
